spec/formatting.md

```diff
@@ -222,6 +222,61 @@ the following steps are taken:
 
    The form that resolved _operand_ and _option_ values take is implementation-defined.
 
+   A _local-declaration_ binds the resolved value of an _expression_
+   to a _variable_.
+   Thus, the output of one _function_ is potentially the _operand_
+   of another _function_. In other words, formatting functions
+   compose with each other.
+   For example, in
+   ```
+   .input {$n :number minIntegerDigits=3}
+   .local {$n1 :number maxFractionDigits=3}
+   ```
+   the second call to `:number` composes with the first call.
+
+   In addition, selector functions compose with formatting functions
+   in the sense that a selector function's _operand_
+   may be the output of any formatting function.
+
+   Implementations SHOULD provide a means for formatting functions
+   to compose with each other
+   and for formatting functions to compose with selector functions.
+   Implementations that provide a means for defining custom functions
+   SHOULD provide a means for those functions to return values
+   that contain enough information
+   (e.g. the resolved _operand_ and _option_ values
+   that the function was called with)
+   to be used as inputs to subsequent function calls.
```

Comment:
This is the key part and I wouldn't want to drop it, as other reviewers suggested. Furthermore, I think we need to say *something somewhere* about resolved values. Otherwise, @aphillips's example from his suggestion above end up being the only place in the spec where we *implicitly* require that resolved values carry some extra information. I'd prefer to be explicit.

Reply:
But we need to avoid being too specific about how it works. "return" has a specific meaning and I don't necessarily think it's a good idea to think of this as a function call's return value.

Our language is a declarative language and we call the stuff in an expression an "annotation" for a reason. Perhaps:

```suggestion
When resolving the value of an _operand_ or other variable
(such as the value of an _option)
implementations SHOULD provide interfaces so that _annotation_
applied in statements can accompany the value where appropriate.
Implementations of _functions_ SHOULD define whether they change the 
value of the _operand_ in any way.
Implementations of _functions_ SHOULD define whether the value of
each _option_ is transitive or local.
```

I mention "statements" here because `.match` can be where the annotation is applied (not just `.local` or `.input`)

Some examples might help here. Here might be an example of non-transitive options (we might say that field options are non-transitive in the spec):

```
.input {$d :datetime weekday=short month=medium day=numeric}
.local $d1 = {$d :datetime hour=|2-digit| minute=numeric}
{{The transaction was on {$d} at {$d1}.}}
```

Here's a similar example:

```
.input {$d :datetime timeZone=|Europe/Paris|}
.local $date = {$d :datetime dateStyle=short}
.local $time = {$date :datetime timeStyle=short}
{{What does {$date} and {$time} print?}}
```

I think it is _less_ surprising if `$time` forgets the earlier style annotation but _not_ the time zone.

We don't currently define any functions that change the value of an operand, but we certainly might do:

```
.local $regular = {|Addison| :string}
.local $shouted = {$regular :transform to=uppercase}
.match {$shouted}
ADDISON {{... is selected... }}
* {{ ... }}
```

Reply:
I see your point about "calling" and "returning".

> (...) so that _annotation_ applied in statements can accompany the value where appropriate.

This sounds OK, although I'm now not sure about the exact meaning of "the value" here. I realize that this was the whole point of @catamorphism's opening the other PR...  

Btw. I think _expressions_ would be more appropriate than _statements_ — it's possible to annotate inside placeholders, too.

> Implementations of _functions_ SHOULD define whether they change the 
> value of the _operand_ in any way.
> Implementations of _functions_ SHOULD define whether the value of
> each _option_ is transitive or local.

Maybe move this part to the note about seeking feedback? We don't really know yet what defining these constraints and requirements should look like.

Reply:
I would also like to avoid using the word "transitive", because transitivity is a property of a mathematical relation and we haven't defined any such relations in the spec.

Reply:
Re: this example:

```
.input {$d :datetime weekday=short month=medium day=numeric}
.local $d1 = {$d :datetime hour=|2-digit| minute=numeric}
{{The transaction was on {$d} at {$d1}.}}
```

It's not obvious to me that the options from the first `:datetime` call shouldn't be preserved -- should `$d1` be a formatted date with the union of all the options shown in both `:datetime` annotations? Or just the `hour` and `minute` options and defaults for the others? To me there's no "obvious" answer, though maybe it's obvious to people who have more experience with message formatting.

It's certainly worth thinking about what the example means, but I'm not sure if it's the best example if the goal is to show something where certain options obviously shouldn't be preserved.

Reply:
On the whole, I understand the suggestion and think it's getting at something useful, but I don't see how it makes sense in the current framework of the spec, in which there is no data model for runtime values. 

I think the first sentence ("...implementations SHOULD provide interfaces so that _annotation_ applied in statements can accompany the value where appropriate.") is too vague, but I'm also not sure how to make it less vague. I would argue that the text in my revised commit is better because it's focused on functions, and the boundary between "inside the formatter" and "in a function implementation" is the one place where information is likely to get "lost". 

I don't get the concept of "changing the value of the operand" even if I mentally replace that with "mapping the operand onto a new operand" or something like that. In your `:transform` example, I would understand `:transform` as returning something like this (if I borrow some of the machinery from #645 but simplify it for presentation):

```
AnnotatedFormattableValue {
  source: AnnotatedFormattableValue { source: Formattable("Addison"), value: FormattedValue("Addison") },
  formatter: "transform",
  options: { "to": "uppercase" },
  value: FormattedValue("ADDISON")
}
```

As with functions in other examples, it passes the same operand through (a wrapped thing ultimately representing the string "Addison"), and the transformed operand (the string "ADDISON") is the "formatted value" thing inside the structure representing the return value.

Obviously this isn't necessarily going to be the data model for runtime values, but still, it's not obvious to me that the result is just "ADDISON" with a bunch of options, rather than something that encapsulates the input "Addison", the output "ADDISON", and the options.

Finally, the sentence about defining options that are or aren't preserved (I would use "preserved" rather than "transitive" as I already said) is important, but we don't have a way in the registry to declare that info, currently.